sql/RFC_SQL.md

```diff
@@ -0,0 +1,1165 @@
+const res = await client.query("SELECT * FROM table");
+await client.close();
+```
+
+Query objects
+
+```ts
+const res = await client.query("SELECT * FROM table");
+console.log(res);
+// [{ col1: "some value" }]
+```
+
+Query one object
+
+```ts
+const res = await client.queryOne("SELECT * FROM table");
+console.log(res);
+// { col1: "some value" }
+```
+
+Query many objects with an iterator
+
+```ts
+const res = Array.fromAsync(client.queryMany("SELECT * FROM table"));
+console.log(res);
+// [{ col1: "some value" }]
+
+// OR
+
+for await (const iterator of client.queryMany("SELECT * FROM table")) {
+  console.log(res);
+  // { col1: "some value" }
+}
+```
+
+Query as an array
+
+```ts
+const res = await client.queryArray("SELECT * FROM table");
+console.log(res);
+// [[ "some value" ]]
+```
+
+Query one as an array
+
+```ts
+const res = await client.queryOneArray("SELECT * FROM table");
+console.log(res);
+// [[ "some value" ]]
+```
+
+Query many as array with an iterator
+
+```ts
+const res = Array.fromAsync(client.queryManyArray("SELECT * FROM table"));
+console.log(res);
+// [[ "some value" ]]
+
+// OR
+
+for await (const iterator of client.queryManyArray("SELECT * FROM table")) {
+  console.log(res);
+  // [ "some value" ]
+}
+```
+
+Query with template literals as an object
+
+```ts
+const res = await client.sql`SELECT * FROM table where id = ${id}`;
+console.log(res);
+// [{ col1: "some value" }]
+```
+
+Query with template literals as an array
+
+```ts
+const res = await client.sqlArray`SELECT * FROM table where id = ${id}`;
+console.log(res);
+// [[ "some value" ]]
+```
+
+Transaction
+
+```ts
+const transaction = await client.beginTransaction();
+await transaction.execute("SOME INSERT QUERY");
+await transaction.commitTransaction();
+// `transaction` can no longer be used, and a new transaction needs to be created
+```
+
+Transaction wrapper
+
+```ts
+const res = await client.transaction(async (t) => {
+  await t.execute("SOME INSERT QUERY");
+  return t.query("SOME SELECT QUERY");
+});
+console.log(res);
+// [{ col1: "some value" }]
+```
+
+Prepared statement
+
+```ts
+const prepared = db.prepare("SOME PREPARED STATEMENT");
+await prepared.query([...params]);
+console.log(res);
+// [{ col1: "some value" }]
+```
+
+## Implementation
+
+> This section is for implementing the interface for database drivers. For
+> general usage, read the [usage](#usage) section.
+
+To be fully compliant with the specs, you will need to implement the following
+classes for your database driver:
+
+- `Connection`: This represents the connection to the database. This should
+  preferably only contain the functionality of containing a connection, and
+  provide a minimum set of query methods to be used to query the database
+- `PreparedStatement`: This represents a prepared statement.
+- `Transaction`: This represents a transaction.
+- `Client`: This represents a database client
+- `ClientPool`: This represents a pool of clients
```

Comment:
How about just `Pool`?

I got confused when I saw `PoolClient` and `ClientPool`.

Or maybe there's not even a need for `PoolClient`, it could simply be a `Client`. 🤷

Reply:
Because the `Client` and the `PoolClient` are not the same and have different options available. 

> How about just Pool?

In this case, i find it better to be descriptive, as it would not immediately tell you if a `Pool` is a `PoolClient` or a `ClientPool`